MimeKit/Cryptography/BouncyCastleSecureMimeContext.cs

```diff
@@ -1300,7 +1233,27 @@ Stream Envelope (CmsRecipientCollection recipients, Stream content, Cancellation
 
 			foreach (var recipient in recipients) {
 				if (unique.Add (recipient.Certificate)) {
-					cms.AddRecipientInfoGenerator (new CmsRecipientInfoGenerator (recipient));
+					var cert = recipient.Certificate;
+					var pub = recipient.Certificate.GetPublicKey();
+					if (pub is RsaKeyParameters) {
+						cms.AddKeyTransRecipient (cert);
+					} else if (pub is ECKeyParameters) {
+						var ecPub = (ECKeyParameters) pub;
+						var kg = GeneratorUtilities.GetKeyPairGenerator ("ECDH");
+						kg.Init(new ECKeyGenerationParameters (ecPub.Parameters, RandomNumberGenerator));
+						var kp = kg.GenerateKeyPair ();
+
+						// TODO: better handle algorithm selection.
+						cms.AddKeyAgreementRecipient (
+							CmsEnvelopedDataGenerator.ECDHSha1Kdf,
+							kp.Public,
+							kp.Private,
+							cert,
+							CmsEnvelopedGenerator.Aes128Wrap
+						);
+					} else {
+						throw new ArgumentException("Unknown type of recipient certificate: " + pub.GetType().Name);
```

Comment:
With a caveat I'm not a native C# developer... and I'm running on Linux... :-D 

[`3e5ab5372`](https://github.com/bcgit/bc-csharp/commit/3e5ab5372) introduced SPKI to X509Certificate and [`c806df63e`](https://github.com/bcgit/bc-csharp/commit/c806df63e) adds some helpers for fetching Algorithm information... I'd probably prefer to surface the OID here rather (or in addition to?) than the BC class name of the key type, as the OID is what the callers of the certificate will have.

Reply:
You mean in the exception? Yea, that probably makes sense.

Reply:
Yeah, in the exception. To do so, I think you need to update BC to a newer version (2.2.1, 2.3.0 is just out which should have this), but not sure how you'd do that from Linux and if you were avoiding it for some compatibility reasons. :-)

Reply:
The other commentary here is that this doesn't handle other types of certs. PQC is coming, and with it brings questions of how to handle their various cert types in places like S/MIME and elsewhere. Personally if let the ink dry on the standards before worrying about that too much here, but IMO, this should set you up a little better for that when the time comes.

Reply:
> To do so, I think you need to update BC to a newer version (I believe 2.2.1 is used in the repo, 2.3.0 is just out which should have this)

I have no problem with bumping BC to 2.3.0 (was planning on doing it anyway). You can actually just edit the MimeKit.csproj and change the version string for BouncyCastle.Cryptography and then run `nuget restore MimeKit.sln`  and/or `dotnet restore MimeKit.sln` and it will download/unpack the version you set in the csproj.

I think `nuget restore` is for .NETFramework and `dotnet restore` works for dotnet core projects. MimeKit is both, so not sure if you need to run both or if one is enough (and if so, which one haha).

Just a useful tip.

Reply:
Perfect, thank you! I've pushed a commit with this information in the error and updating BC to 2.3.0.